### PR TITLE
fix(templating): append and join steps should template dereferenced pipelines

### PR DESCRIPTION
**Pull request by rhuille**
https://trello.com/c/WbYGUUxA/4308-as-a-conceptor-i-could-use-some-templating-on-my-right-dataset-when-im-using-the-join-vqb-step

### CODE REVIEW

File: src/lib/steps.ts [hunk 1]
```diff
@@ -6,6 +6,10 @@ type PrimitiveType = number | boolean | string | Date;
 type Templatable<T> = T | string;
 export type Reference = Pipeline | string;
 
+export function isNotDeReferenced(pipelineOrReference: Reference): pipelineOrReference is string {
```
Comment (davinov) [inline src/lib/steps.ts:9]: Maybe `isReference` would a simpler name that avoids the double negation

File: src/lib/templating.ts [hunk 2]
```diff
@@ -263,7 +271,11 @@ export class PipelineInterpolator implements StepMatcher<S.PipelineStep> {
   }
 
   join(step: Readonly<S.JoinStep>) {
-    return { ...step };
+    if (S.isNotDeReferenced(step.right_pipeline)) {
+      return { ...step };
+    } else {
+      return { ...step, right_pipeline: this.interpolate(step.right_pipeline) };
+    }
```
Comment (davinov) [inline src/lib/templating.ts:278]: Is there a case when we arrive here and the pipeline is still referenced?
Maybe for validation when we submit the step form?

I suggest we indicate when this happens, so further readers will understand why the two cases are necessary.
(and if not... remove one :D)

**Review by davinov — COMMENTED** [review]
Thanks, here a two questions:

File: src/lib/steps.ts (hunk 1 above)
Comment (rhuille) [reply src/lib/steps.ts]: https://github.com/ToucanToco/weaverbird/pull/654/commits/b5e00576c0ba6322cb27e9e757ea083c5f26ec0a

File: src/lib/templating.ts (hunk 2 above)
Comment (rhuille) [reply src/lib/templating.ts]: Bonne idée! https://github.com/ToucanToco/weaverbird/pull/654/commits/7891b0e4be90b70dafe86daf14f6817b06c05065

**Comment by davinov** [comment]
@rhuille does this solve https://github.com/ToucanToco/weaverbird/issues/564 ? :)

**Comment by rhuille** [comment]
Tout à fait ! :)

**Comment by rhuille** [comment]
Tout à fait ! :)

**Review by davinov — APPROVED** [review]
Thanks!

**Pull request merged**